upgrade/steps.go

```diff
@@ -805,3 +805,54 @@ func ReplaceAssertNoError(ctx Context, repo ProviderRepo) (step.Step, error) {
 	}
 	return step.Combined("Replace AssertNoError with AssertNoErrorf", steps...), nil
 }
+
+// UpgradePulumiVersions reads the current Pulumi SDK version from go.mod and applies it to:
+// sdk/go.mod
+// examples/go.mod - we also infer the `pkg` version here and add it.
+func BridgePulumiVersions(ctx Context, repo ProviderRepo) (step.Step, error) {
+	// When we've updated the bridge version, we need to update the corresponding pulumi version in sdk/go.mod.
+	// It needs to match the version used in provider/go.mod, which is *not* necessarily `latest`.
+	var newSdkVersion string
+	steps := []step.Step{}
+	getNewPulumiVersionStep := step.F("Get Pulumi SDK version", func() (string, error) {
+		modFile := filepath.Join(repo.root, "provider", "go.mod")
+		lookupModule := "github.com/pulumi/pulumi/sdk/v3"
+		pulumiMod, found, err := currentGoVersionOf(modFile, lookupModule)
+		if err != nil {
+			return "", err
+		}
+		if !found {
+			return "", fmt.Errorf("%s: %s not found\n", modFile, lookupModule)
+		}
+		return pulumiMod.Version, nil
+
+	}).AssignTo(&newSdkVersion)
+
+	bridgePulumiVersionSdkStep := step.Computed(func() step.Step {
+		return step.Cmd(exec.CommandContext(ctx,
+			"go",
+			"get",
+			"github.com/pulumi/pulumi/sdk/v3@"+newSdkVersion)).In(repo.sdkDir())
+	})
+	bridgePulumiVersionExamplesStep := step.Computed(func() step.Step {
+		return step.Cmd(exec.CommandContext(ctx,
+			"go",
+			"get",
+			"github.com/pulumi/pulumi/sdk/v3@"+newSdkVersion)).In(repo.examplesDir())
+	})
+	// In examples/ folder, we use pulumi/pkg as well. We should keep this in sync also.
+	examplesPulumiPkgStep := step.Computed(func() step.Step {
+		return step.Cmd(exec.CommandContext(ctx,
+			"go",
+			"get",
+			"github.com/pulumi/pulumi/pkg/v3@"+newSdkVersion)).In(repo.examplesDir())
+	})
+
+	steps = append(steps,
+		getNewPulumiVersionStep,
+		bridgePulumiVersionSdkStep,
+		bridgePulumiVersionExamplesStep,
+		examplesPulumiPkgStep)
```

Comment:
I find this much easier to parse, since we make it clear that each upgrade is the same, parameterized over the mod path:
```suggestion
	upgrade := func(mod string) step.Step {
		return step.Computed(func() step.Step {
			return step.Cmd(exec.CommandContext(ctx,
				"go", "get",
				"github.com/pulumi/pulumi/"+mod+"/v3@"+newSdkVersion))
		})
	}

	steps = append(steps,
		getNewPulumiVersionStep,
		upgrade("sdk").In(repo.sdkDir()),
		upgrade("sdk").In(repo.examplesDir()),
		// In examples/ folder, we use pulumi/pkg as well. We should keep this in sync also.
		upgrade("pkg").In(repo.examplesDir()))
```

This is what I meant by moving repetitive code into a function.

Reply:
Yes, I tried something like that. And again, I cannot figure out how to make steps.Comupted() do the right thing - the `newSDKVersion` doesn't get evaluated.